kratos/utilities/sensitivity_builder.h

```diff
@@ -0,0 +1,282 @@
+            mNodalSolutionStepSensitivityVariables[i] = nodal_sensitivity_variables[i].GetString();
+        }
+
+        Parameters element_sensitivity_variables =
+            Settings["element_data_sensitivity_variables"];
+        mElementDataSensitivityVariables.resize(element_sensitivity_variables.size());
+        for (unsigned int i = 0; i < element_sensitivity_variables.size(); ++i)
+        {
+            mElementDataSensitivityVariables[i] =
+                element_sensitivity_variables[i].GetString();
+        }
+
+        mBuildMode = Settings["build_mode"].GetString();
+
+        KRATOS_CATCH("");
+    }
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    void Initialize()
+    {
+        KRATOS_TRY;
+
+        SetAllSensitivityVariablesToZero();
+
+        VariableUtils().SetNonHistoricalVariable(UPDATE_SENSITIVITIES, false,
+                                                 mrModelPart.Nodes());
+        VariableUtils().SetNonHistoricalVariable(
+            UPDATE_SENSITIVITIES, true, mpSensitivityModelPart->Nodes());
+
+        KRATOS_CATCH("");
+    }
+
+    void UpdateSensitivities()
+    {
+        KRATOS_TRY;
+
+        double weight;
+        if (mBuildMode == "integrate")
+        {
+            // integrate in time.
+            weight = -mrModelPart.GetProcessInfo()[DELTA_TIME];
+        }
+        else if (mBuildMode == "sum")
+        {
+            // sum in time.
+            weight = 1.0;
+        }
+        else if (mBuildMode == "static")
+        {
+            // overwrite existing.
+            SetAllSensitivityVariablesToZero();
+            weight = 1.0;
+        }
+        else
+        {
+            KRATOS_ERROR << "Unsupported \"build_mode\": " << mBuildMode << std::endl;
+        }
+
+        NodalSensitivityBuilder nodal_sensitivity_builder(mrModelPart, mpResponseFunction);
+        for (const std::string& r_label : mNodalSolutionStepSensitivityVariables)
+        {
+            nodal_sensitivity_builder.BuildNodalSolutionStepSensitivities(r_label, weight);
+        }
+        
+        ElementSensitivityBuilder element_sensitivity_builder(mrModelPart, mpResponseFunction);
+        for (const std::string& r_label : mElementDataSensitivityVariables)
+        {
+            element_sensitivity_builder.BuildElementSensitivities(r_label, weight);
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    ///@}
+
+protected:
+    ///@name Protected member Variables
+    ///@{
+
+    ///@}
+    ///@name Protected Operators
+    ///@{
+
+    ///@}
+    ///@name Protected Operations
+    ///@{
+
+    ///@}
+
+private:
+    ///@name Member Variables
+    ///@{
+
+    ModelPart& mrModelPart;
+    ModelPart* mpSensitivityModelPart = nullptr;
+    AdjointResponseFunction::Pointer mpResponseFunction;
+    std::vector<std::string> mNodalSolutionStepSensitivityVariables;
+    std::vector<std::string> mElementDataSensitivityVariables;
+    std::string mBuildMode;
```

Comment:
Probably an `enum` would make sense for `mBuildMode`.

Reply:
I would keep it as string since it is less complexity with negligible cost difference.